berteome
================

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->

## Install

`pip install berteome`

# How to use

Berteome makes use of the masked language model of BERT to determine
predictions for all residues in a protein sequence.

The main `berteome` library can be imported as follows:

``` python
from berteome import berteome
```

The `modelLoader` class can be used to show what models are supported by
`berteome`.

``` python
berteome_models = berteome.modelLoader()
```

``` python
berteome_models.supported_models
```

    ['Rostlab/prot_bert',
     'facebook/esm2_t33_650M_UR50D',
     'facebook/esm1b_t33_650M_UR50S']

All of these models are distributed through huggingface, and berteome
makes great use of it’s API.

## Load library

To load prot_bert model, run the following:

``` python
bert_tokenizer, bert_model = berteome_models.load_model("Rostlab/prot_bert")
```

    Downloading:   0%|          | 0.00/81.0 [00:00<?, ?B/s]

    Downloading:   0%|          | 0.00/112 [00:00<?, ?B/s]

    Downloading:   0%|          | 0.00/86.0 [00:00<?, ?B/s]

    Downloading:   0%|          | 0.00/361 [00:00<?, ?B/s]

    Downloading:   0%|          | 0.00/1.68G [00:00<?, ?B/s]

    Some weights of the model checkpoint at Rostlab/prot_bert were not used when initializing BertForMaskedLM: ['cls.seq_relationship.weight', 'cls.seq_relationship.bias']
    - This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
    - This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).

The language models utilized by `berteome` were trained using a masked
token approach. In this approach, a random amino acid is masked in a
protein and the model is trained to predict what the amino acid should
be. These models do this on an incredibly large amount of protein
sequences, to the point that they begin to learn the language of protein
sequence space as we currently know it. For instance, it can start to
learn, which residues are unlikely to exist at a given point in a
protein. Using these models, you can place a mask at any given residue
in the protein, and the model will generate a probability score for all
the possible amino acids that could go there.

`berteome` allows the user to take the models and begin to really
investigate these predictions for a given protein sequence, by masking
every single residue in the protein sequence and predicting the
probabilities for all the possible amino acids. The result is a nice,
easy to work with pandas data frame. To make this dataframe for a very
simple peptide sequence (`MENDEL`), do the following:

``` python
mendel_bertDF = berteome.modelPredDF("MENDEL",bert_tokenizer, bert_model)
mendel_bertDF.predDf
```

  <div id="df-4622814e-007d-4581-888a-a5505bda2bf7">
    <div class="colab-df-container">
      <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>wt</th>
      <th>wtIndex</th>
      <th>wtScore</th>
      <th>A</th>
      <th>C</th>
      <th>D</th>
      <th>E</th>
      <th>F</th>
      <th>G</th>
      <th>H</th>
      <th>...</th>
      <th>M</th>
      <th>N</th>
      <th>P</th>
      <th>Q</th>
      <th>R</th>
      <th>S</th>
      <th>T</th>
      <th>V</th>
      <th>W</th>
      <th>Y</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>M</td>
      <td>1</td>
      <td>0.076602</td>
      <td>0.036697</td>
      <td>0.011504</td>
      <td>0.048245</td>
      <td>0.118906</td>
      <td>0.024072</td>
      <td>0.039202</td>
      <td>0.012621</td>
      <td>...</td>
      <td>0.076602</td>
      <td>0.072661</td>
      <td>0.024722</td>
      <td>0.038672</td>
      <td>0.043105</td>
      <td>0.070280</td>
      <td>0.056544</td>
      <td>0.049927</td>
      <td>0.007781</td>
      <td>0.021699</td>
    </tr>
    <tr>
      <th>1</th>
      <td>E</td>
      <td>2</td>
      <td>0.074830</td>
      <td>0.045721</td>
      <td>0.015662</td>
      <td>0.041921</td>
      <td>0.074830</td>
      <td>0.037153</td>
      <td>0.044325</td>
      <td>0.018264</td>
      <td>...</td>
      <td>0.043581</td>
      <td>0.062667</td>
      <td>0.025277</td>
      <td>0.036911</td>
      <td>0.055543</td>
      <td>0.064425</td>
      <td>0.049955</td>
      <td>0.056789</td>
      <td>0.012691</td>
      <td>0.029893</td>
    </tr>
    <tr>
      <th>2</th>
      <td>N</td>
      <td>3</td>
      <td>0.041990</td>
      <td>0.043564</td>
      <td>0.009685</td>
      <td>0.162590</td>
      <td>0.184364</td>
      <td>0.033782</td>
      <td>0.044661</td>
      <td>0.012355</td>
      <td>...</td>
      <td>0.041484</td>
      <td>0.041990</td>
      <td>0.019992</td>
      <td>0.025515</td>
      <td>0.029433</td>
      <td>0.048106</td>
      <td>0.030303</td>
      <td>0.054742</td>
      <td>0.007430</td>
      <td>0.024924</td>
    </tr>
    <tr>
      <th>3</th>
      <td>D</td>
      <td>4</td>
      <td>0.049748</td>
      <td>0.042083</td>
      <td>0.013244</td>
      <td>0.049748</td>
      <td>0.086194</td>
      <td>0.039736</td>
      <td>0.055911</td>
      <td>0.016861</td>
      <td>...</td>
      <td>0.040080</td>
      <td>0.060822</td>
      <td>0.032024</td>
      <td>0.039689</td>
      <td>0.046228</td>
      <td>0.062323</td>
      <td>0.044901</td>
      <td>0.058937</td>
      <td>0.010875</td>
      <td>0.026596</td>
    </tr>
    <tr>
      <th>4</th>
      <td>E</td>
      <td>5</td>
      <td>0.086915</td>
      <td>0.046641</td>
      <td>0.018770</td>
      <td>0.079822</td>
      <td>0.086915</td>
      <td>0.050638</td>
      <td>0.050466</td>
      <td>0.022397</td>
      <td>...</td>
      <td>0.028962</td>
      <td>0.062234</td>
      <td>0.023879</td>
      <td>0.030534</td>
      <td>0.040489</td>
      <td>0.065195</td>
      <td>0.044938</td>
      <td>0.068038</td>
      <td>0.012156</td>
      <td>0.038034</td>
    </tr>
    <tr>
      <th>5</th>
      <td>L</td>
      <td>6</td>
      <td>0.060736</td>
      <td>0.038191</td>
      <td>0.009217</td>
      <td>0.065189</td>
      <td>0.152547</td>
      <td>0.020950</td>
      <td>0.049525</td>
      <td>0.013955</td>
      <td>...</td>
      <td>0.040042</td>
      <td>0.096484</td>
      <td>0.020712</td>
      <td>0.035022</td>
      <td>0.046888</td>
      <td>0.049071</td>
      <td>0.046247</td>
      <td>0.048276</td>
      <td>0.010486</td>
      <td>0.022727</td>
    </tr>
  </tbody>
</table>
<p>6 rows × 23 columns</p>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-4622814e-007d-4581-888a-a5505bda2bf7')"
              title="Convert this dataframe to an interactive table."
              style="display:none;">
        
  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"/>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"/><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"/>
  </svg>
      </button>
      
  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-4622814e-007d-4581-888a-a5505bda2bf7 button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-4622814e-007d-4581-888a-a5505bda2bf7');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
  

This dataframe is where the true berteomic magic begins. Each row
corresponds to each residue in the input protein sequence.

The `wt` column represents the actual amino acid at the given position,
and `wtIndex` is just a one-based index of the residue which makes
plotting easier, may not stick around forever though..

`wtScore` is a very interesting and important value. For a given
protein, one would hope that the model would predict that the masked
residue would be the same as the wild-type in the sequence. This column
gives us the actual probability that the model provided for the wild
type residue at that position.

The remaining columns are simply the probabilities of each possible
amino acid generated by the model when placing a mask at every residue
in the input protein.

Using `MENDEL` as a toy example, isn’t truly a great use casage, because
the model wasn’t necesarily trained on such small peptides with that
particular order of residues, nevertheless, it still demonstrates the
possibilities that `berteome` opens.

From 6 amino acids, we have 120 data points to start playing around
with!

From a single sequence, you could generate a seqlogo plot, which would
normally require aligning multiple homologous sequences together.

A position-specific scoring matrix for BLAST searches and HMM searches
could be generated from a single sequence!

From a single sequence, you could also generate a sampling of variants
that would be reasonably better than just randomly mutating, allowing
for a decent way of mutating a protein sequence or just augmenting a
sparse protein dataset!

# Development

To build the library run the following

    nbdev export

Then, pip install in a development environment

    pip install -e '.[dev]'

I do quite a bit of work on a chromebook, which allows for doing stuff
on github through codespace and also on google colab. To install a
particular commit hash of `berteome` you can do the following:

``` python
!pip uninstall berteome
```

    Found existing installation: berteome 0.1.5
    Uninstalling berteome-0.1.5:
      Would remove:
        /usr/local/lib/python3.8/dist-packages/berteome-0.1.5.dist-info/*
        /usr/local/lib/python3.8/dist-packages/berteome/*
    Proceed (y/n)? y
      Successfully uninstalled berteome-0.1.5

``` python
!pip install "berteome @ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4"
```

    Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
    Collecting berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4
      Cloning https://github.com/tijeco/berteome (to revision 96c830219a42353732c2745916dd262f5f6b36f4) to /tmp/pip-install-7aytir0f/berteome_12e6af4a03da4db5be231813fd882354
      Running command git clone -q https://github.com/tijeco/berteome /tmp/pip-install-7aytir0f/berteome_12e6af4a03da4db5be231813fd882354
      Running command git rev-parse -q --verify 'sha^96c830219a42353732c2745916dd262f5f6b36f4'
      Running command git fetch -q https://github.com/tijeco/berteome 96c830219a42353732c2745916dd262f5f6b36f4
      Running command git checkout -q 96c830219a42353732c2745916dd262f5f6b36f4
    Requirement already satisfied: pip in /usr/local/lib/python3.8/dist-packages (from berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (21.1.3)
    Requirement already satisfied: packaging in /usr/local/lib/python3.8/dist-packages (from berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (21.3)
    Requirement already satisfied: pandas in /usr/local/lib/python3.8/dist-packages (from berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (1.3.5)
    Requirement already satisfied: numpy in /usr/local/lib/python3.8/dist-packages (from berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (1.21.6)
    Requirement already satisfied: seqlogo in /usr/local/lib/python3.8/dist-packages (from berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (5.29.8)
    Requirement already satisfied: transformers in /usr/local/lib/python3.8/dist-packages (from berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (4.25.1)
    Requirement already satisfied: torch in /usr/local/lib/python3.8/dist-packages (from berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (1.13.0+cu116)
    Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /usr/local/lib/python3.8/dist-packages (from packaging->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (3.0.9)
    Requirement already satisfied: python-dateutil>=2.7.3 in /usr/local/lib/python3.8/dist-packages (from pandas->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (2.8.2)
    Requirement already satisfied: pytz>=2017.3 in /usr/local/lib/python3.8/dist-packages (from pandas->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (2022.6)
    Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.8/dist-packages (from python-dateutil>=2.7.3->pandas->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (1.15.0)
    Requirement already satisfied: weblogo in /usr/local/lib/python3.8/dist-packages (from seqlogo->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (3.7.12)
    Requirement already satisfied: ghostscript in /usr/local/lib/python3.8/dist-packages (from seqlogo->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (0.7)
    Requirement already satisfied: pytest in /usr/local/lib/python3.8/dist-packages (from seqlogo->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (3.6.4)
    Requirement already satisfied: setuptools>=38.6.0 in /usr/local/lib/python3.8/dist-packages (from ghostscript->seqlogo->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (57.4.0)
    Requirement already satisfied: more-itertools>=4.0.0 in /usr/local/lib/python3.8/dist-packages (from pytest->seqlogo->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (9.0.0)
    Requirement already satisfied: attrs>=17.4.0 in /usr/local/lib/python3.8/dist-packages (from pytest->seqlogo->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (22.1.0)
    Requirement already satisfied: atomicwrites>=1.0 in /usr/local/lib/python3.8/dist-packages (from pytest->seqlogo->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (1.4.1)
    Requirement already satisfied: pluggy<0.8,>=0.5 in /usr/local/lib/python3.8/dist-packages (from pytest->seqlogo->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (0.7.1)
    Requirement already satisfied: py>=1.5.0 in /usr/local/lib/python3.8/dist-packages (from pytest->seqlogo->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (1.11.0)
    Requirement already satisfied: typing-extensions in /usr/local/lib/python3.8/dist-packages (from torch->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (4.4.0)
    Requirement already satisfied: huggingface-hub<1.0,>=0.10.0 in /usr/local/lib/python3.8/dist-packages (from transformers->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (0.11.1)
    Requirement already satisfied: filelock in /usr/local/lib/python3.8/dist-packages (from transformers->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (3.8.2)
    Requirement already satisfied: requests in /usr/local/lib/python3.8/dist-packages (from transformers->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (2.23.0)
    Requirement already satisfied: pyyaml>=5.1 in /usr/local/lib/python3.8/dist-packages (from transformers->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (6.0)
    Requirement already satisfied: tokenizers!=0.11.3,<0.14,>=0.11.1 in /usr/local/lib/python3.8/dist-packages (from transformers->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (0.13.2)
    Requirement already satisfied: regex!=2019.12.17 in /usr/local/lib/python3.8/dist-packages (from transformers->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (2022.6.2)
    Requirement already satisfied: tqdm>=4.27 in /usr/local/lib/python3.8/dist-packages (from transformers->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (4.64.1)
    Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.8/dist-packages (from requests->transformers->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (3.0.4)
    Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.8/dist-packages (from requests->transformers->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (1.24.3)
    Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.8/dist-packages (from requests->transformers->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (2022.12.7)
    Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.8/dist-packages (from requests->transformers->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (2.10)
    Requirement already satisfied: scipy in /usr/local/lib/python3.8/dist-packages (from weblogo->seqlogo->berteome@ git+https://github.com/tijeco/berteome@96c830219a42353732c2745916dd262f5f6b36f4) (1.7.3)
    Building wheels for collected packages: berteome
      Building wheel for berteome (setup.py) ... done
      Created wheel for berteome: filename=berteome-0.1.5-py3-none-any.whl size=12836 sha256=2e48134dcab49582081ac37d1ba26e95f57a075d8b97a7e10b6b46bbb15ef785
      Stored in directory: /root/.cache/pip/wheels/67/50/32/3e870391d4962242696251e34f45ce7e2a7eb5a37a2a41c6df
    Successfully built berteome
    Installing collected packages: berteome
    Successfully installed berteome-0.1.5

    Unable to display output for mime type(s): application/vnd.colab-display-data+json
