{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alright, so I think I have a pretty neat idea for this package, so I'm just going to document a few of my thoughts. This is an augment package, where the idea is to take a protein sequence and generate a bunch of point mutational variants. I think this would be quite handy in certain applications, for instance antimicrobial peptide classification. There aren't too many known examples of antimicrobial peptides, but tons and tons of examples of non-antimicrobial peptides. If this were a computer vision problem, say we were trying to build a classifier for four leaf clovers vs three leaf clovers, you'd have a much easier time finding examples of three leaf vs four leaf. To compensate for this so that you have an evenish sampling you can augment the dataset byt manipulating the pictures you do have and jittering the images about to get kind of pseudoreplicates. Sure, there's lot's of ways to approach that problem, but for protein sequences it is not so easy to augment those sequences. If they are all in the same gene family, you can make a multiple sequence alignment of them and reconstruct their ancestral states using phylogenetic approaches, this would double the data size! But if that isn't the case then what can you do? Well you can take the single point mutation approach. Say for a protein of length L, generate N variants where you randomly mutate a single amino residue. That is sort of fine in practice, but we know that some mutations are more feleterious than others, if only there were a way to determine which mutations weren't likely to be deleterious... Aha! So I think we can take the dataframes generated from ESM or prot_bert and generate the top k most likely mutational variants per residue. Let's think about how that could potentially augment the data set. So for a single peptide of length L, if we take the top k substitutions per residue, that gives L*k variants, for the toy MENDEL example where L is 6, this would generate 30 variants! Which is quite substantial, for larger proteins I think you could get quite a reasanobly large number of variants.\n",
    "\n",
    "Alright, so here is what I plan to do, the prediction dataframe is pretty core to this, though for demonstration purposes, I'll probably just handmake a MENDEL dataframe. From that the main function will be an augmentPep function which will take a prediction df from a single peptide, as well k to dictate how many mutational variants per residue. An obvious caveat is the fact that the wild type residue will probably often be in the top k possible substitutions, so there will need to be a small check to insure we are getting the top k residues that are not wt. Beyond that, two aditional functions come to mind than can be built on top of this, augmentFasta and augmentPeps, which just takes augmentPep and applies it to either a fasta file or an iterable list of peptides. I should also think about how to name mutants, I feel like just posResidue + ResidueSub would  be sufficient to tag that to the original id.\n",
    "\n",
    "Let's try to jump in this, and make a dummy MENDEL dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>F</th>\n",
       "      <th>G</th>\n",
       "      <th>H</th>\n",
       "      <th>I</th>\n",
       "      <th>K</th>\n",
       "      <th>...</th>\n",
       "      <th>M</th>\n",
       "      <th>N</th>\n",
       "      <th>P</th>\n",
       "      <th>Q</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>V</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  wt  A  C  D  E  F  G  H  I  K  ...  M  N  P  Q  R  S  T  V  W  Y\n",
       "0  M  0  0  0  0  0  0  0  0  0  ...  9  7  0  0  0  0  0  0  0  0\n",
       "1  E  0  0  7  8  9  0  0  0  0  ...  0  0  0  0  0  0  0  0  0  0\n",
       "2  N  0  0  0  0  0  0  0  0  0  ...  8  7  0  0  0  0  0  0  0  0\n",
       "3  D  0  7  8  9  0  0  0  0  0  ...  0  0  0  0  0  0  0  0  0  0\n",
       "4  L  0  0  0  0  0  0  0  0  8  ...  7  0  0  0  0  0  0  0  0  0\n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mendelDF = pd.DataFrame.from_dict({\n",
    "\t#\t A,C,D,E,F,G,H,I,K,L,M,N,P,Q,R,S,T,V,W,Y\n",
    "\t\"M\":[0,0,0,0,0,0,0,0,0,8,9,7,0,0,0,0,0,0,0,0],\n",
    "\t\"E\":[0,0,7,8,9,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],\n",
    "\t\"N\":[0,0,0,0,0,0,0,0,0,9,8,7,0,0,0,0,0,0,0,0],\n",
    "\t\"D\":[0,7,8,9,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],\n",
    "\t# \"E\":[0,0,9,8,7,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0], # redundant... for the example\n",
    "\t\"L\":[0,0,0,0,0,0,0,0,8,9,7,0,0,0,0,0,0,0,0,0]\n",
    "}, orient=\"index\", columns=list(\"ACDEFGHIKLMNPQRSTVWY\")).reset_index().rename(columns={\"index\": \"wt\"})\n",
    "\n",
    "\n",
    "mendelDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alright so this very much a barebones dummy df, but I think it will work.\n",
    "\n",
    "So for this example I would only want to sample k=2, to go through each residue it should be something like\n",
    "\n",
    "1. M: L,N\n",
    "2. E: D,F\n",
    "3. N: L,M\n",
    "4. D: C,E\n",
    "<!-- 5. E: D,F -->\n",
    "6. L: K,M\n",
    "\n",
    "So that should give us 12 variant sequences! (Edit:10..)\n",
    "\n",
    "So let me go ahead and think about what the logic needs to be for this. There is certainly a chance that I will be able to use some fancy pandas stuff to do this, but for now I will just focus on what might work.\n",
    "\n",
    "So I definitely know that I need to grab the top k values in each column.. wait that's not right. No, I need to get the top values for each row! Then the tricky part is figuring out if the column name is the same value as the index.\n",
    "\n",
    "\n",
    "Oh gravy, there definitely has to be some sort of SQLy or panday way to get this\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'0xL': 'LENDL',\n",
       " '0xN': 'NENDL',\n",
       " '1xF': 'MFNDL',\n",
       " '1xD': 'MDNDL',\n",
       " '2xL': 'MELDL',\n",
       " '2xM': 'MEMDL',\n",
       " '3xE': 'MENEL',\n",
       " '3xC': 'MENCL',\n",
       " '4xK': 'MENDK',\n",
       " '4xM': 'MENDM'}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def augmentPep(df, k):\n",
    "\tseqList = list(df[\"wt\"])\n",
    "\tvariantDict = {}\n",
    "\tfor index, row in df.iterrows():\n",
    "\t\tscores = row[list(\"ACDEFGHIKLMNPQRSTVWY\")]\n",
    "\t\ttop_k_scores = scores.where(scores.index != row[\"wt\"]).sort_values(ascending=False).head(k)\n",
    "\n",
    "\t\ttop_k_subs = list(top_k_scores.index)\n",
    "\t\tfor res in top_k_subs:\n",
    "\t\t\t\n",
    "\t\t\tseqCopy = seqList.copy()\n",
    "\t\t\tseqCopy[index] = res\n",
    "\t\t\tvariantDict[f\"{index}x{res}\"] = ''.join(seqCopy) \n",
    "\n",
    "\treturn variantDict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert augmentPep(mendelDF, 2) == {'0xL': 'LENDL',\n",
    " '0xN': 'NENDL',\n",
    " '1xF': 'MFNDL',\n",
    " '1xD': 'MDNDL',\n",
    " '2xL': 'MELDL',\n",
    " '2xM': 'MEMDL',\n",
    " '3xE': 'MENEL',\n",
    " '3xC': 'MENCL',\n",
    " '4xK': 'MENDK',\n",
    " '4xM': 'MENDM'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ALright!! In principle, that's working!!\n",
    "\n",
    "The output structure needs to be modified so that the mutational variant is annotated as \"pos_res\".\n",
    "\n",
    "I think I like this..\n",
    "\n",
    "Let's see how it works on esm/prot_bert dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Some weights of the model checkpoint at Rostlab/prot_bert were not used when initializing BertForMaskedLM: ['cls.seq_relationship.weight', 'cls.seq_relationship.bias']\n",
      "- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).\n",
      "- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).\n"
     ]
    }
   ],
   "source": [
    "from berteome import prot_bert"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mendel_prot_bert_DF = prot_bert.bertPredictionDF(\"MENDEL\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt</th>\n",
       "      <th>wtIndex</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>F</th>\n",
       "      <th>G</th>\n",
       "      <th>H</th>\n",
       "      <th>I</th>\n",
       "      <th>...</th>\n",
       "      <th>M</th>\n",
       "      <th>N</th>\n",
       "      <th>P</th>\n",
       "      <th>Q</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>V</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>0.036685</td>\n",
       "      <td>0.011501</td>\n",
       "      <td>0.048229</td>\n",
       "      <td>0.118868</td>\n",
       "      <td>0.024064</td>\n",
       "      <td>0.039190</td>\n",
       "      <td>0.012617</td>\n",
       "      <td>0.066477</td>\n",
       "      <td>...</td>\n",
       "      <td>0.076580</td>\n",
       "      <td>0.072637</td>\n",
       "      <td>0.024714</td>\n",
       "      <td>0.038660</td>\n",
       "      <td>0.043091</td>\n",
       "      <td>0.070257</td>\n",
       "      <td>0.056526</td>\n",
       "      <td>0.049911</td>\n",
       "      <td>0.007779</td>\n",
       "      <td>0.021692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>2</td>\n",
       "      <td>0.045712</td>\n",
       "      <td>0.015659</td>\n",
       "      <td>0.041913</td>\n",
       "      <td>0.074816</td>\n",
       "      <td>0.037146</td>\n",
       "      <td>0.044317</td>\n",
       "      <td>0.018260</td>\n",
       "      <td>0.073063</td>\n",
       "      <td>...</td>\n",
       "      <td>0.043572</td>\n",
       "      <td>0.062655</td>\n",
       "      <td>0.025272</td>\n",
       "      <td>0.036905</td>\n",
       "      <td>0.055532</td>\n",
       "      <td>0.064412</td>\n",
       "      <td>0.049945</td>\n",
       "      <td>0.056779</td>\n",
       "      <td>0.012689</td>\n",
       "      <td>0.029887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>0.043558</td>\n",
       "      <td>0.009684</td>\n",
       "      <td>0.162566</td>\n",
       "      <td>0.184336</td>\n",
       "      <td>0.033777</td>\n",
       "      <td>0.044654</td>\n",
       "      <td>0.012353</td>\n",
       "      <td>0.052622</td>\n",
       "      <td>...</td>\n",
       "      <td>0.041478</td>\n",
       "      <td>0.041984</td>\n",
       "      <td>0.019989</td>\n",
       "      <td>0.025511</td>\n",
       "      <td>0.029428</td>\n",
       "      <td>0.048098</td>\n",
       "      <td>0.030299</td>\n",
       "      <td>0.054734</td>\n",
       "      <td>0.007428</td>\n",
       "      <td>0.024920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D</td>\n",
       "      <td>4</td>\n",
       "      <td>0.042079</td>\n",
       "      <td>0.013243</td>\n",
       "      <td>0.049744</td>\n",
       "      <td>0.086189</td>\n",
       "      <td>0.039733</td>\n",
       "      <td>0.055907</td>\n",
       "      <td>0.016860</td>\n",
       "      <td>0.073291</td>\n",
       "      <td>...</td>\n",
       "      <td>0.040078</td>\n",
       "      <td>0.060817</td>\n",
       "      <td>0.032022</td>\n",
       "      <td>0.039686</td>\n",
       "      <td>0.046224</td>\n",
       "      <td>0.062319</td>\n",
       "      <td>0.044898</td>\n",
       "      <td>0.058933</td>\n",
       "      <td>0.010875</td>\n",
       "      <td>0.026594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>5</td>\n",
       "      <td>0.046638</td>\n",
       "      <td>0.018769</td>\n",
       "      <td>0.079816</td>\n",
       "      <td>0.086908</td>\n",
       "      <td>0.050634</td>\n",
       "      <td>0.050462</td>\n",
       "      <td>0.022395</td>\n",
       "      <td>0.074495</td>\n",
       "      <td>...</td>\n",
       "      <td>0.028960</td>\n",
       "      <td>0.062229</td>\n",
       "      <td>0.023877</td>\n",
       "      <td>0.030532</td>\n",
       "      <td>0.040486</td>\n",
       "      <td>0.065190</td>\n",
       "      <td>0.044934</td>\n",
       "      <td>0.068032</td>\n",
       "      <td>0.012155</td>\n",
       "      <td>0.038031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>L</td>\n",
       "      <td>6</td>\n",
       "      <td>0.035695</td>\n",
       "      <td>0.008615</td>\n",
       "      <td>0.060928</td>\n",
       "      <td>0.142576</td>\n",
       "      <td>0.019581</td>\n",
       "      <td>0.046287</td>\n",
       "      <td>0.013043</td>\n",
       "      <td>0.060374</td>\n",
       "      <td>...</td>\n",
       "      <td>0.037424</td>\n",
       "      <td>0.090177</td>\n",
       "      <td>0.019358</td>\n",
       "      <td>0.032733</td>\n",
       "      <td>0.043823</td>\n",
       "      <td>0.045863</td>\n",
       "      <td>0.043224</td>\n",
       "      <td>0.045121</td>\n",
       "      <td>0.009800</td>\n",
       "      <td>0.021241</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  wt  wtIndex         A         C         D         E         F         G  \\\n",
       "0  M        1  0.036685  0.011501  0.048229  0.118868  0.024064  0.039190   \n",
       "1  E        2  0.045712  0.015659  0.041913  0.074816  0.037146  0.044317   \n",
       "2  N        3  0.043558  0.009684  0.162566  0.184336  0.033777  0.044654   \n",
       "3  D        4  0.042079  0.013243  0.049744  0.086189  0.039733  0.055907   \n",
       "4  E        5  0.046638  0.018769  0.079816  0.086908  0.050634  0.050462   \n",
       "5  L        6  0.035695  0.008615  0.060928  0.142576  0.019581  0.046287   \n",
       "\n",
       "          H         I  ...         M         N         P         Q         R  \\\n",
       "0  0.012617  0.066477  ...  0.076580  0.072637  0.024714  0.038660  0.043091   \n",
       "1  0.018260  0.073063  ...  0.043572  0.062655  0.025272  0.036905  0.055532   \n",
       "2  0.012353  0.052622  ...  0.041478  0.041984  0.019989  0.025511  0.029428   \n",
       "3  0.016860  0.073291  ...  0.040078  0.060817  0.032022  0.039686  0.046224   \n",
       "4  0.022395  0.074495  ...  0.028960  0.062229  0.023877  0.030532  0.040486   \n",
       "5  0.013043  0.060374  ...  0.037424  0.090177  0.019358  0.032733  0.043823   \n",
       "\n",
       "          S         T         V         W         Y  \n",
       "0  0.070257  0.056526  0.049911  0.007779  0.021692  \n",
       "1  0.064412  0.049945  0.056779  0.012689  0.029887  \n",
       "2  0.048098  0.030299  0.054734  0.007428  0.024920  \n",
       "3  0.062319  0.044898  0.058933  0.010875  0.026594  \n",
       "4  0.065190  0.044934  0.068032  0.012155  0.038031  \n",
       "5  0.045863  0.043224  0.045121  0.009800  0.021241  \n",
       "\n",
       "[6 rows x 22 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mendel_prot_bert_DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'0xE': 'EENDEL',\n",
       " '0xK': 'KENDEL',\n",
       " '1xL': 'MLNDEL',\n",
       " '1xK': 'MKNDEL',\n",
       " '2xE': 'MEEDEL',\n",
       " '2xD': 'MEDDEL',\n",
       " '3xL': 'MENLEL',\n",
       " '3xK': 'MENKEL',\n",
       " '4xL': 'MENDLL',\n",
       " '4xD': 'MENDDL',\n",
       " '5xE': 'MENDEE',\n",
       " '5xK': 'MENDEK'}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "augmentPep(mendel_prot_bert_DF, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from berteome import esm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mendel_esm_DF = esm.esmPredictionDF(\"MENDEL\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'0xE': 'EENDEL',\n",
       " '0xD': 'DENDEL',\n",
       " '1xL': 'MLNDEL',\n",
       " '1xS': 'MSNDEL',\n",
       " '2xE': 'MEEDEL',\n",
       " '2xL': 'MELDEL',\n",
       " '3xL': 'MENLEL',\n",
       " '3xK': 'MENKEL',\n",
       " '4xL': 'MENDLL',\n",
       " '4xS': 'MENDSL',\n",
       " '5xE': 'MENDEE',\n",
       " '5xK': 'MENDEK'}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "augmentPep(mendel_esm_DF, 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As far as I'm concerned, this is working! It shouldn't be too difficult to use this to make a `augmentPeps` and `augmentFasta`, which takes multiple peptides and returns their top k variants.\n",
    "\n",
    "I don't entirely need this right this moment, so maybe I should just work on that when I definitely need it or at least have more time, because it will probably take a couple of hours to finish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.3 64-bit",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
